### PR TITLE
Use HTTPHeaderDict in request

### PR DESCRIPTION
**Pull request by sigmavirus24**
Supersedes #633

### CODE REVIEW

**Comment by sigmavirus24** [comment]
This might not be perfect, but it at least passes the `test_collections` tests. And it's late here =P

**Comment by sigmavirus24** [comment]
Tests are failing because there isn't that tests the `get` method directly of the `HTTPHeaderDict`.

File: urllib3/_collections.py [hunk 1]
```diff
@@ -290,33 +294,35 @@ def copy(self):
...
         """Iterate over all headers, merging duplicate ones together."""
         for key in self:
-            val = _dict_getitem(self, key)
+            val = self._container[key.lower()]
             yield val[0], ', '.join(val[1:])
 
     def items(self):
         return list(self.iteritems())
 
+    viewitems = items
```
Comment (sigmavirus24) [inline urllib3/_collections.py]: Forgot to delete this. We don't need it.

**Comment by Lukasa** [comment]
Broadly LGTM.

**Comment by shazow** [comment]
Mmmm yea this is a smaller change than I thought we'd need. If you can make the remaining test(s) pass, I'm all for it. :)

**Comment by shazow** [comment]
Btw, :custard:.

**Comment by sigmavirus24** [comment]
So the tests *should* pass now. The thing that I'm uncertain about is [this test](https://github.com/shazow/urllib3/commit/8d7273d97997c34d0f22c06fad7d96312f2db536#diff-1ec38ab7dbd425bce37c385e1e250b95R241). I can definitely see that `dict(HTTPHeaderDict(baz='quuz')) == {'baz': 'quuz'}` but something else is turning that into `{'Baz': 'quuz'}`. I haven't investigated it yet.

**Comment by shazow** [comment]
Hm wonder if it's our dummyserver and/or httplib mangling headers on some versions.

**Comment by sigmavirus24** [comment]
> Hm wonder if it's our dummyserver and/or httplib mangling headers on some versions.

It's more bizarre because this test *was* passing before we fixed the round-trip behaviour. =/

**Comment by sigmavirus24** [comment]
If I debug into the request methods, I get to 

```py
(Pdb) p kw['headers']
{'Accept': '*/*', 'baz': 'quux', 'Host': 'localhost:52261'}
```

On L276 in `urllib3/poolmanager.py`

**Comment by sigmavirus24** [comment]
If I go further, it's still the right case in `connectionpool.py`. I don't have much time to go further though. Sorry. =(

**Comment by sigmavirus24** [comment]
For anyone else looking to follow the same rabbit hole.

1. Add `import pdb; pdb.set_trace()` (or `ipdb` if that's your thing) to `test/with_dummyserver/test_proxy_poolmanager.py` at L238.
2. `tox -e py27 -- -s test.with_dummyserver.test_proxy_poolmanager` (or `py34` or whatever)
3. Start stepping into functions that make the request.

**Comment by Lukasa** [comment]
So this isn't in `httplib`: the header is emitted in the case it was provided to `httplib`. Current money is on the dummyserver doing that.

**Comment by Lukasa** [comment]
Definitely the dummy server. From tcpdump:

```
GET http://localhost:62953/headers HTTP/1.1
Accept-Encoding: identity
Accept: */*
Host: localhost:62953
baz: quux
foo: bar

HTTP/1.1 200 OK
Etag: "feb15a7ce2766d24cf20cf1633197fc5d8a59281"
Date: Sun, 19 Jul 2015 10:22:50 GMT
Content-Length: 146
Server: TornadoServer/4.1
Content-Type: text/html; charset=UTF-8,text/plain

{"Accept": "*/*", "Content-Length": "0", "Foo": "bar", "Connection": "close", "Host": "localhost:62953", "Accept-Encoding": "gzip", "Baz": "quux"}
```

We definitely emitted headers in lowercase, they were definitely reflected to us in upper-case.

**Comment by Lukasa** [comment]
Note also that the dummyserver inserts some headers we didn't send (`Connection: close`) and mangles one we did send (`Accept-Encoding: gzip`).

**Comment by Lukasa** [comment]
Ok, so had a quick chat with @shazow and I think that test should just be rethought. We can't effectively test that case is preserved when we round-trip via the dummy server, so let's just not.

I can see a case for wanting a test that ensures that case is preserved that includes httplib, but if we're going to do that we should do it as a socket-level test where we can confirm that the bytes-on-the-wire are really what we think they are.

**Comment by sigmavirus24** [comment]
So b49490d has a socketlevel test for this.

**Comment by shazow** [comment]
Yay :Custard:!

Are we good to go? Shall I merge and ship v1.11?

**Comment by sigmavirus24** [comment]
That's Cory's call.
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by Lukasa** [comment]
Let's rock-and-roll. :shipit:

**Comment by mmedvede** [comment]
This causes OpenStack devstack-gate tests to fail on Fedora 21 (but not on Ubuntu). There are interactions between many python packages during the test, so I am not yet certain if it is urlib's fault, but if I use the commit previous to the merge of this pull request, tempest tests pass.

**Comment by Lukasa** [comment]
@mmedvede Can we get more information, e.g. a link to a Launchpad bug?

**Comment by sigmavirus24** [comment]
@Lukasa it appears they're tracking it in https://bugs.launchpad.net/glance/+bug/1476770. The original bug was because the gate is not operating as it is supposed to (and I suspect the same is happening on Fedora). The gate should **always** be installing requests from pip instead of using system packages but they're not so when they install urllib3 1.11.0, it installs over the system urllib3 that was installed. I maintain that this is fundamentally a problem with the gate. Note that instead of fixing the gate, they released a new version of oslo.vmware (the only library to use urllib3 in openstack directly) with a cap to exclude 1.11.0. This is not a problem or a bug in urllib3.

**Comment by mmedvede** [comment]
@sigmavirus24 I was not aware of the bug, thanks. It is exactly what was happening to our third-party Fedora 21 gate (IBM PowerKVM CI) before I disabled the urllib3 1.11 install. I assume that Ubuntu installs newer system python-requests, so when urllib3 1.11 is installed on top, the gate doesn't break. The weird thing is that both Ubuntu and Fedora gates show requests 2.7.0 installed. I have confirmed that doing 'pip install --upgrade --force-reinstall requests' fixes Fedora 21 in my scenario, even though requests would still show as being 2.7.0.

@Lukasa Thanks for quick response. Not urllib3 problem.

**Comment by sigmavirus24** [comment]
@mmedvede check where requests 2.7.0 is installed from. You'll see that it's installed from the system packages and that is *not* what the gate is supposed to do. Anyway, further discussion should happen on LaunchPad, not here.

**Comment by tfmorris** [comment]
This change requires all `requests-cache` caches be deleted.

**Pull request merged**